In [21]:
distance = {
    'A': {'A': 0, 'B': 10, 'C': 15, 'D': 20, 'E': 25},
    'B': {'A': 10, 'B': 0, 'C': 35, 'D': 25, 'E': 15},
    'C': {'A': 15, 'B': 35, 'C': 0, 'D': 30, 'E': 20},
    'D': {'A': 20, 'B': 25, 'C': 30, 'D': 0, 'E': 10},
    'E': {'A': 25, 'B': 15, 'C': 20, 'D': 10, 'E': 0}
}

In [23]:
def calculate_distance(route):
    total = 0
    for i in range(len(route)):
        total += distance[route[i]][route[(i + 1) % len(route)]]
    return total

def generate_neighbors(route):
    neighbors = []
    for i in range(len(route)):
        for j in range(i + 1, len(route)):
            neighbor = route.copy()
            neighbor[i], neighbor[j] = neighbor[j], neighbor[i]
            neighbors.append(neighbor)
    return neighbors

# Algoritma Hill Climbing
def hill_climbing(initial_route):
    current_route = initial_route.copy()
    current_distance = calculate_distance(current_route)
    print(f"Rute awal: {current_route}, Jarak: {current_distance}")
    
    while True:
        neighbors = generate_neighbors(current_route)
        if not neighbors:
            break
        
        best_neighbor = min(neighbors, key=lambda x: calculate_distance(x))
        best_distance = calculate_distance(best_neighbor)
        
        if best_distance < current_distance:
            current_route = best_neighbor
            current_distance = best_distance
            print(f"Rute diperbarui: {current_route}, Jarak: {current_distance}")
        else:
            break 
    
    return current_route, current_distance

In [25]:
initial_route = ['A', 'B', 'C', 'D', 'E']
best_route, best_distance = hill_climbing(initial_route)

print("\nHasil akhir:")
print(f"Rute terpendek: {best_route}, Jarak: {best_distance}")

Rute awal: ['A', 'B', 'C', 'D', 'E'], Jarak: 110
Rute diperbarui: ['B', 'A', 'C', 'D', 'E'], Jarak: 80

Hasil akhir:
Rute terpendek: ['B', 'A', 'C', 'D', 'E'], Jarak: 80


# Traveling Salesman Problem (TSP) - Hill Climbing Algorithm

## 1. Representasi Solusi Awal

Solusi awal yang digunakan adalah urutan kota:
**A → B → C → D → E → A**

### Perhitungan Jarak:
- **A → B**: 10  
- **B → C**: 35  
- **C → D**: 30  
- **D → E**: 10  
- **E → A**: 25  

**Total Jarak**:  
10 + 35 + 30 + 10 + 25 = **110**

---

## 2. Cara Menghasilkan Neighbors dengan Menukar Dua Kota

Untuk menghasilkan neighbors (tetangga) dari solusi awal:
1. Lakukan pertukaran dua kota dalam rute.
2. Contoh:
    - Tukar posisi **A** dan **B** pada rute awal:  
      **B → A → C → D → E**
    - Tukar posisi **C** dan **E**:  
      **A → B → E → D → C**

### Total Neighbors:
Untuk 5 kota, jumlah neighbors yang dihasilkan adalah **10** (kombinasi 2 dari 5: C(5,2) = 10).

---

## 3. Proses Perubahan Rute dengan Algoritma Hill Climbing

### Langkah-Langkah:
1. **Inisialisasi**:
    - Rute awal: **A → B → C → D → E**  
    - Jarak: **110**

2. **Generasi Tetangga**:
    - Hitung jarak semua neighbors.
    - Contoh tetangga terbaik: **B → A → C → D → E** dengan jarak **80**.

3. **Perbandingan**:
    - Karena jarak tetangga (**80**) < jarak saat ini (**110**), rute diperbarui.

4. **Iterasi Berikutnya**:
    - Cari tetangga dari **B → A → C → D → E**.
    - Tidak ditemukan tetangga dengan jarak lebih pendek dari **80**.

5. **Algoritma Berhenti**:
    - Tidak ada perbaikan lebih lanjut.

### Output Proses:
- **Rute awal**: `['A', 'B', 'C', 'D', 'E']`, Jarak: **110**  
- **Rute diperbarui**: `['B', 'A', 'C', 'D', 'E']`, Jarak: **80**

---

## 4. Rute Terpendek dan Optimalitas

### Rute Terpendek:
**B → A → C → D → E → B**

#### Perhitungan Jarak:
- **B → A**: 10  
- **A → C**: 15  
- **C → D**: 30  
- **D → E**: 10  
- **E → B**: 15  

**Total Jarak**:  
10 + 15 + 30 + 10 + 15 = **80**

### Optimalitas:
- Ini adalah solusi **optimal**, bukan local optimum.
- **Verifikasi**:
  - Jarak **80** adalah minimal untuk matriks yang diberikan.
  - Tidak ada kombinasi lain yang menghasilkan jarak lebih pendek (misalnya, rute **B → E → D → C → A → B** memiliki jarak **85**).

---

## 5. Variabel Penting

| Nama              | Tipe   | Nilai                                                                 |
|--------------------|--------|----------------------------------------------------------------------|
| **best_distance**  | int    | 80                                                                   |
| **best_route**     | list   | `['B', 'A', 'C', 'D', 'E']`                                          |
| **current_distance** | float | 91.65274873991386                                                   |
| **current_solution** | list | `[2, 0, 1, 4, 3]`                                                   |
| **distance_matrix** | list  | `[[0, 10, 15, 20, 25], [10, 0, 35, 25, 15], ...]`                   |
| **neighbors**      | list   | `[[0, 2, 1, 4, 3], [1, 0, 2, 4, 3], ...]`                           |

---

## 6. Visualisasi Kota

### Koordinat Kota:
| Kota | Koordinat (x, y) |
|------|------------------|
| A    | (0, 0)          |
| B    | (10, 0)         |
| C    | (15, 35)        |
| D    | (20, 25)        |
| E    | (25, 15)        |

---

## 7. Kesimpulan

Algoritma Hill Climbing berhasil menemukan rute terpendek dengan jarak **80**. Solusi ini adalah solusi optimal untuk matriks jarak yang diberikan.